url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/util/Discounts.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+package edu.columbia.stat.wood.pub.sequencememoizer.util;
+
+import java.io.Serializable;
+
+/**
+ * Discount parameter wrapper for sequence memoizer.
+ * 
+ * @author nicholasbartlett
+ */
+public class Discounts implements Serializable{
+
+    static final long serialVersionUID = 1;
+
+    private double[] discounts, logDiscounts, discountGradient;
+    private double alpha, alphaGradient;
+
+    /**
+     * Sets the discount parameters.
+     *
+     * @param initialDiscounts unique discount parameters
+     * @param dInfinity discount between restaurant of depth (initialDiscounts.length -1) and restaurants infinitely deep
+     */
+    public Discounts(double[] initialDiscounts, double dInfinity){
+        if(dInfinity <= 0.0 || dInfinity >= 1.0){
+            throw new IllegalArgumentException("dInfinity must be in the interval (0.0,1.0)");
+        }
+
+        alpha = Math.log(dInfinity) / (Math.log(dInfinity) + Math.log(initialDiscounts[initialDiscounts.length-1]));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;initialDiscounts&#39; (index: 0 | type: double[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;dInfinity&#39;. The same parameter type is validated here for example: url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/IntSequenceMemoizerParameters.java:33. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).